Analysis/include/EventInfo.h

```diff
@@ -195,44 +233,107 @@ class EventInfoBase {
+            if(selected_signal_jets.isSelectedBjet(n)) continue;
+            if(selected_signal_jets.isSelectedVBFjet(n)) continue;
+            double tag;
+            if(jet_ordering == JetOrdering::Pt)
+                tag = event.jets_p4.at(n).Pt();
+            else if(jet_ordering == JetOrdering::CSV)
+                tag = event.jets_csv.at(n);
+            else if(jet_ordering == JetOrdering::DeepCSV)
+                tag = event.jets_deepCsv_BvsAll.at(n);
+            else
+                throw exception("Unsupported jet ordering for b-jet pair selection.");
+            jet_info_vector_new.emplace_back(event.jets_p4.at(n),n,tag);
+        }
+
+        auto new_bjets_ordered = jet_ordering::OrderJets(jet_info_vector_new,true,bjet_pt_cut,bjet_eta_cut);
+        if(new_bjets_ordered.size() >= 1)
+            selected_signal_jets.selectedBjetPair.second = new_bjets_ordered.at(0).index;
+        else{
+            selected_signal_jets.selectedVBFjetPair = ntuple::UndefinedJetPair();
+            if (bjets_ordered.size() >= 2)
+                selected_signal_jets.selectedBjetPair.second = bjets_ordered.at(1).index;
+        }
+
+        return selected_signal_jets;
     }
 
     std::array<size_t,2> GetSelectedBjetIndices() const
     {
         std::array<size_t,2> bjet_indexes;
-        bjet_indexes[0] = selected_bjet_pair.first;
-        bjet_indexes[1] = selected_bjet_pair.second;
+        bjet_indexes[0] = selected_signal_jets.selectedBjetPair.first;
+        bjet_indexes[1] = selected_signal_jets.selectedBjetPair.second;
         return bjet_indexes;
     }
 
     std::set<size_t> GetSelectedBjetIndicesSet() const
     {
         std::set<size_t> bjet_indexes;
-        bjet_indexes.insert(selected_bjet_pair.first);
-        bjet_indexes.insert(selected_bjet_pair.second);
+        bjet_indexes.insert(selected_signal_jets.selectedBjetPair.first);
+        bjet_indexes.insert(selected_signal_jets.selectedBjetPair.second);
         return bjet_indexes;
     }
 
 
     static constexpr int verbosity = 0;
 
-    EventInfoBase(const Event& _event, const JetPair& _selected_bjet_pair = JetPair(0, 1),
+    EventInfoBase(const Event& _event, const analysis::Period& _period, JetOrdering _jet_ordering,
                   const SummaryInfo* _summaryInfo = nullptr) :
         event(&_event), summaryInfo(_summaryInfo), eventIdentifier(_event.run, _event.lumi, _event.evt),
-        selected_bjet_pair(_selected_bjet_pair),
-        has_bjet_pair(selected_bjet_pair.first < GetNJets() && selected_bjet_pair.second < GetNJets())
+        selected_signal_jets(SelectSignalJets(_event,_period,_jet_ordering)), period(_period),
+        jet_ordering(_jet_ordering)
     {
         mutex = std::make_shared<Mutex>();
         triggerResults.SetAcceptBits(event->trigger_accepts);
         triggerResults.SetMatchBits(event->trigger_matches);
+        kinfitProducer = std::shared_ptr<analysis::kin_fit::FitProducer>(new analysis::kin_fit::FitProducer());
```

Comment:
it is not efficient to create kinfit producer for each event. Use static method.